platforms/software/dependency-management/src/main/java/org/gradle/api/internal/artifacts/ivyservice/resolveengine/artifact/DefaultResolvedArtifactsBuilder.java

```diff
@@ -57,6 +57,7 @@ private void collectArtifacts(int artifactSetId, ArtifactSet artifacts) {
     }
 
     public VisitedArtifactResults complete() {
-        return new DefaultVisitedArtifactResults(Collections.unmodifiableList(artifactSetsById));
+        // Copy to shrink the list to the actual size
+        return new DefaultVisitedArtifactResults(ImmutableList.copyOf(artifactSetsById));
```

Comment:
Yeah not sure _how_ much this helps, but the idea is to shrink the array to the proper size.

When we are appending to it above while visiting a resolution's artifacts, we continuously double the array size (as an array list does).

This change creates a new list with the proper size and passes that along.

This is important since this list is retained for the entire duration of a Gradle sync (and an entire build without CC). So The length of this list is (roughly) equal to the number of nodes in a graph. Multiply that by the number of resolvable configurations in a project. Multiply that by the number of projects in a build.